balloon/src/lib.rs

```diff
@@ -0,0 +1,347 @@
+        salt: &[u8],
+        memory_blocks: &mut [GenericArray<u8, D::OutputSize>],
+    ) -> Result<GenericArray<u8, D::OutputSize>> {
+        if self.params.p_cost.get() > 1 {
+            return Err(Error::ThreadsTooMany);
+        }
+
+        let s_cost = usize::try_from(self.params.s_cost.get()).unwrap();
+        let t_cost = usize::try_from(self.params.t_cost.get()).unwrap();
+
+        let mut digest = D::new();
+
+        // This is a direct translation of the `Balloon` from <https://eprint.iacr.org/2016/027.pdf> chapter 3.1.
+        // int delta = 3 // Number of dependencies per block
+        const DELTA: u32 = 3;
+        // int cnt = 0 // A counter (used in security proof)
+        let mut cnt: u32 = 0;
+        // block_t buf[s_cost]): // The main buffer
+        let memory_blocks = memory_blocks
+            .get_mut(..s_cost)
+            .ok_or(Error::MemoryTooLittle)?;
+
+        // Step 1. Expand input into buffer.
+        // buf[0] = hash(cnt++, passwd, salt)
+        digest.update(cnt.to_le_bytes());
+        cnt = cnt.wrapping_add(1);
+        digest.update(pwd);
+        digest.update(salt);
+
+        if let Some(secret) = self.secret {
+            digest.update(secret);
+        }
+
+        memory_blocks[0] = digest.finalize_reset();
+
+        // for m from 1 to s_cost-1:
+        for m in 1..s_cost {
+            // buf[m] = hash(cnt++, buf[m-1])
+            digest.update(&cnt.to_le_bytes());
+            cnt = cnt.wrapping_add(1);
+            digest.update(&memory_blocks[m - 1]);
+            memory_blocks[m] = digest.finalize_reset();
+        }
+
+        // Step 2. Mix buffer contents.
+        // for t from 0 to t_cost-1:
+        for t in 0..t_cost {
+            // for m from 0 to s_cost-1:
+            for m in 0..s_cost {
+                // Step 2a. Hash last and current blocks.
+                // block_t prev = buf[(m-1) mod s_cost]
+                let prev = if m == 0 { s_cost - 1 } else { m - 1 };
+                let prev = &memory_blocks[prev];
+
+                // buf[m] = hash(cnt++, prev, buf[m])
+                digest.update(&cnt.to_le_bytes());
+                cnt = cnt.wrapping_add(1);
+                digest.update(prev);
+                digest.update(&memory_blocks[m]);
+                memory_blocks[m] = digest.finalize_reset();
+
+                // Step 2b. Hash in pseudorandomly chosen blocks.
+                // for i from 0 to delta-1:
+                for i in 0..DELTA {
+                    // block_t idx_block = ints_to_block(t, m, i)
+                    // int other = to_int(hash(cnt++, salt, idx_block)) mod s_cost
+                    digest.update(&cnt.to_le_bytes());
+                    cnt = cnt.wrapping_add(1);
+                    digest.update(salt);
+                    digest.update(&u32::try_from(t).unwrap().to_le_bytes());
+                    digest.update(&u32::try_from(m).unwrap().to_le_bytes());
+                    digest.update(&i.to_le_bytes());
```

Comment:
I am using `u32` here, which is arbitrary, as far as I could see the research paper says nothing about it. The reference implementation uses `i32` I believe, other implementations out there use `u32` and `u64`.

Reply:
I think `cnt` and `i` should both be `u64`. If you had a 5GHz computer that did an increment every clock cycle it'd take over 100 years to overflow. Whereas with a `u32` it'd take less than a second.

Reply:
I changed `cnt` to be u64 to prevent overflow, but `i` can only reach 3, as big as `delta` basically. So protecting it from overflow won't be necessary. In any case, I had the brilliant idea to change everything into `u64`.